device-telemetry/app/com/microsoft/azure/iotsolutions/devicetelemetry/services/storage/timeSeries/TimeSeriesClient.java

```diff
@@ -0,0 +1,244 @@
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import play.Logger;
+import play.libs.Json;
+import play.libs.ws.WSClient;
+import play.libs.ws.WSRequest;
+import play.libs.ws.WSResponse;
+
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+public class TimeSeriesClient implements ITimeSeriesClient {
+
+    private static final Logger.ALogger log = Logger.of(TimeSeriesConfig.class);
+    public static final int CLOCK_CALIBRATION_IN_SECONDS = 5;
+
+    private final WSClient wsClient;
+
+    private final String AUTHORITY = "https://login.windows.net/%s";
+    private final String RESOURCE = "https://api.timeseries.azure.com/";
+
+    private final String AVAILABILITY_KEY = "availability";
+    private final String EVENTS_KEY = "events";
+    private final String SEARCH_SPAN_KEY = "searchSpan";
+    private final String PREDICATE_KEY = "predicate";
+    private final String PREDICATE_STRING_KEY = "predicateString";
+    private final String TOP_KEY = "top";
+    private final String SORT_KEY = "sort";
+    private final String SORT_INPUT_KEY = "input";
+    private final String BUILT_IN_PROP_KEY = "builtInProperty";
+    private final String BUILT_IN_PROP_VALUE = "$ts";
+    private final String SORT_ORDER_KEY = "order";
+    private final String COUNT_KEY = "count";
+    private final String FROM_KEY = "from";
+    private final String TO_KEY = "to";
+    private final String DEVICE_ID_KEY = "iothub-connection-device-id";
+
+
+    private String aadTenantId;
+    private String applicationId;
+    private String applicationSecret;
+    private String dataAccessFqdn;
+    private String apiVersion;
+    private String dateFormat;
+    private int timeoutInSeconds;
+
+    private static AuthenticationResult authenticationResult;
+
+    @Inject
+    public TimeSeriesClient(IServicesConfig config, WSClient wsClient)
+        throws InvalidConfigurationException {
+        this.wsClient = wsClient;
+
+        MessagesConfig messagesConfig = config.getMessagesConfig();
+        if (messagesConfig == null) {
+            throw new InvalidConfigurationException("Message config is empty");
+        }
+
+        TimeSeriesConfig tsiConfig = messagesConfig.getTimeSeriesConfig();
+
+        if (tsiConfig != null) {
+            this.dataAccessFqdn = tsiConfig.getTimeSeriesFqdn();
+            this.aadTenantId = tsiConfig.getAadTenant();
+            this.applicationId = tsiConfig.getAadApplicationId();
+            this.applicationSecret = tsiConfig.getAadApplicationSecret();
+            this.apiVersion = tsiConfig.getApiVersion();
+            this.dateFormat = tsiConfig.getDateFormat();
+            this.timeoutInSeconds = tsiConfig.getTimeOutInSeconds();
+        } else {
+            log.info("Time Series config is empty");
```

Comment:
Should we throw instead?

Reply:
When CosmosDB is enabled instead of TSI, the DI is still try to instantiate Message with TimeSeriesClient. To avoid null point exception, we have to return it. We will rely on the storage type to determine which storage to query.